init.lua

```diff
@@ -1,77 +1,302 @@
+local mod_name = "wielded_light"
+
 local update_interval = 0.2
+local cleanup_interval = update_interval*3
+local velocity_projection = update_interval * 1
 local level_delta = 2
+local max_light_level = 14
 local shiny_items = {}
 
---- Shining API ---
-wielded_light = {}
+local active_lights = {}
+local light_recalcs = {}
+local tracked_entities = {}
+local update_callbacks = {}
+local update_player_callbacks = {}
 
-function wielded_light.update_light(pos, light_level)
-	local around_vector = {
-			{x=0, y=0, z=0},
-			{x=0, y=1, z=0}, {x=0, y=-1, z=0},
-			{x=1, y=0, z=0}, {x=-1, y=0, z=0},
-			{x=0, y=0, z=1}, {x=0, y=0, z=1},
-		}
-	local do_update = false
-	local old_value = 0
-	local timer
-	local light_pos
-	for _, around in ipairs(around_vector) do
-		light_pos = vector.add(pos, around)
-		local name = minetest.get_node(light_pos).name
-		if name == "air" and (minetest.get_node_light(light_pos) or 0) < light_level then
-			do_update = true
-			break
-		elseif name:sub(1,13) == "wielded_light" then -- Update existing light node and timer
-			old_value = tonumber(name:sub(15))
-			if light_level > old_value then
-				do_update = true
-			else
-				timer = minetest.get_node_timer(light_pos)
-				local elapsed = timer:get_elapsed()
-				if elapsed > (update_interval * 1.5) then
-					-- The timer is set to 3x update_interval
-					-- This node was not updated the last interval and may
-					-- is disabled before the next step
-					-- Therefore the light should be re-set to avoid flicker
-					do_update = true
+local cat_id = 0
+local cat_codes = {}
+local function get_cat_code(cat)
```

Comment:
Rename suggestion: `get_light_category_id()`.

More importantly: Why do you need categories?
Whatever ID it outputs - it's not reliable between server restarts when a mod is enabled to adds (or makes use of) another category.

Reply:
It doesn't need to be reliable over restarts, the entities being referred to will also change their IDs so all the old lights should get removed and new ones created.

Categories are required so you can assign multiple wielded lights to the same entity without getting them confused - this is because of a new item I am trying to add that acts like a "mining helmet", basically a helmet that glows so you can mine without placing torches down or requiring an "off-hand".
This concept brings in the the question of what happens when you try to impose two wielded light values (on the head and in the hand) on the same space in different globalstep loops - it flickers between them.
Having categories allows you to keep any number (..well, up to 256.. any number within reason) of wielded light items on the same entity at a time, if you wanted a full set of glow armor, a torch in one hand and glow glass in the other, you could have a set of categories, one for each armor position and one for each hand. This makes it possible to keep track of what light source is where when otherwise they would get lost in whatever process you are using to sum up the light levels, making the light level unreliable.

Reply:
Renamed

Reply:
Excuse me, but why couldn't we just use the category name directly? That's already an unique identifier for the type of light.

Reply:
I did add a comment at the top of this function that explains that

Reply:
```
	Using 2-digit hex codes for categories
	Starts at 00, ends at FF
	This makes it easier extract `uid` from `cat_id..uid` by slicing off 2 characters
	The category ID must be of a fixed length (2 characters)
```